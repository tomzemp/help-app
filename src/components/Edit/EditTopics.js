import { Button, InputField } from '@dhis2/ui'
import PropTypes from 'prop-types'
import React, { useState } from 'react'
import { connect } from 'react-redux'
import {
    thuAddTopicToSummaryEdit,
    thuReorderTopicsEditing,
} from '../../actions/settingsEdit'
import { IconMoveDown, IconMoveUp } from '../../icons/icons'
import i18n from '../../locales'
import { selGetTopicsEdit } from '../../reducers/settingsEdit'

const TopicsList = ({ currentTopics, reorderTopics }) => {
    const [selectedID, setSelectedID] = useState('')

    return (
        <>
            <div className="outerItemListBox">
                <div className="itemListBox">
                    <div className="innerItemListBox">
                        {currentTopics.map(t => (
                            <TopicListItem
                                id={t.id}
                                key={`option_${t.name}`}
                                name={t.name}
                                selected={t.id === selectedID}
                                setSelectedID={setSelectedID}
                            />
                        ))}
                    </div>
                    <div className="actionButtonsContainer">
                        <div className="actionButtons">
                            <div className="actionButton">
                                <Button
                                    disabled={
                                        !selectedID ||
                                        currentTopics
                                            .map(t => t.id)
                                            .indexOf(selectedID) === 0
                                    }
                                    icon={<IconMoveUp />}
                                    key="button_move_up"
                                    name="up_button"
                                    onClick={() => {
                                        reorderTopics(selectedID, 'up')
                                    }}
                                    small
                                    type="button"
                                    value="up"
                                ></Button>
                            </div>
                            <div className="actionButton">
                                <Button
                                    disabled={
                                        !selectedID ||
                                        currentTopics
                                            .map(t => t.id)
                                            .indexOf(selectedID) ===
                                            currentTopics.length - 1
                                    }
                                    icon={<IconMoveDown />}
                                    key="button_move_down"
                                    name="down_button"
                                    onClick={() => {
                                        reorderTopics(selectedID, 'down')
                                    }}
                                    small
                                    type="button"
                                    value="down"
                                ></Button>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
            <style jsx>{`
                .outerItemListBox {
                    width: 500px;
                    margin-bottom: var(--spacers-dp16);
                }
                .itemListBox {
                    margin-right: var(--spacers-dp4);
                    border: 1px solid var(--colors-grey500);
                    border-radius: 3px;
                }
                .innerItemListBox {
                    min-height: 200px;
                    max-height: 300px;
                    overflow-y: auto;
                }
                .actionButtonsContainer {
                    border-top: 1px solid var(--colors-grey500);
                    padding: var(--spacers-dp4);
                    display: flex;
                }
                .actionButtons {
                    margin-left: auto;
                    display: flex;
                }
                .actionButton {
                    margin-left: var(--spacers-dp4);
                }
            `}</style>
        </>
    )
}

TopicsList.propTypes = {
    currentTopics: PropTypes.array,
    reorderTopics: PropTypes.func,
}

const TopicListItem = ({ id, name, selected, setSelectedID }) => (
    <>
        <div
            className={selected ? 'topicListDivHighlighted' : 'topicListDiv'}
            onClick={() => {
                setSelectedID(id)
            }}
            id={id}
        >
            <span key={`span_${id}`}>{name}</span>
        </div>
        <style jsx>{`
            .topicListDiv {
                overflow: hidden;
                padding: var(--spacers-dp4);
                cursor: default;
            }
            .topicListDivHighlighted {
                overflow: hidden;
                padding: var(--spacers-dp4);
                background-color: var(--colors-teal700);
                cursor: default;
            }
            .topicListDivHighlighted span {
                color: var(--colors-white);
            }
            .topicListDiv:hover {
                background-color: var(--colors-grey200);
            }
        `}</style>
    </>
)

TopicListItem.propTypes = {
    id: PropTypes.string,
    name: PropTypes.string,
    selected: PropTypes.bool,
    setSelectedID: PropTypes.func,
}

const AddNewTopic = ({
    addTopic,
    checkIfTopicExists,
    newTopicName,
    setNewTopicName,
}) => (
    <>
        <div>
            <InputField
                error={checkIfTopicExists(newTopicName)}
                inputWidth="300px"
                onChange={target => {
                    setNewTopicName(target.value)
                }}
                placeholder={i18n.t('Enter a new topic name')}
                value={newTopicName}
                validationText={
                    checkIfTopicExists(newTopicName)
                        ? i18n.t('Topic name already exists')
                        : ''
                }
            />
            <div className="addButton">
                <Button
                    disabled={
                        !newTopicName.length || checkIfTopicExists(newTopicName)
                    }
                    onClick={() => {
                        addTopic(newTopicName)
                        setNewTopicName('')
                    }}
                    primary
                    type="button"
                >
                    {i18n.t('Add Topic')}
                </Button>
            </div>
        </div>
        <style jsx>{`
            .addButton {
                margin-top: var(--spacers-dp4);
            }
        `}</style>
    </>
)

AddNewTopic.propTypes = {
    addTopic: PropTypes.func,
    checkIfTopicExists: PropTypes.func,
    newTopicName: PropTypes.string,
    setNewTopicName: PropTypes.func,
}

const EditTopics = ({ addTopic, reorderTopics, topics }) => {
    const [newTopicName, setNewTopicName] = useState('')

    const checkIfTopicExists = name =>
        Boolean(topics.find(t => t.name === name))

    return (
        <>
            <TopicsList
                currentTopics={topics}
                reorderTopics={reorderTopics}
            ></TopicsList>
            <AddNewTopic
                addTopic={addTopic}
                checkIfTopicExists={checkIfTopicExists}
                newTopicName={newTopicName}
                setNewTopicName={setNewTopicName}
            />
        </>
    )
}

const mapStateToProps = state => {
    return {
        topics: selGetTopicsEdit(state),
    }
}

const mapDispatchToProps = {
    addTopic: thuAddTopicToSummaryEdit,
    reorderTopics: thuReorderTopicsEditing,
}

EditTopics.propTypes = {
    addTopic: PropTypes.func,
    reorderTopics: PropTypes.func,
    topics: PropTypes.array,
}

export default connect(mapStateToProps, mapDispatchToProps)(EditTopics)
