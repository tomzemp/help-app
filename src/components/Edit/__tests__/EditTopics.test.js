import React from 'react'
import { render } from '../../../testing/test-utils'
import '@testing-library/jest-dom/extend-expect'
import EditTopics from '../EditTopics'

describe('EditTopics', () => {
    it('Renders as expected', () => {
        const { container } = render(<EditTopics />, {
            initialState: {
                settingsEdit: {
                    topics: [
                        { id: 'r', name: 'rabbits' },
                        { id: 'd', name: 'dogs' },
                    ],
                },
            },
        })
        expect(container).toMatchSnapshot()
    })
})
