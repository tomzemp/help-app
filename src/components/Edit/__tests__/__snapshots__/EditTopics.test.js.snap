// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditTopics Renders as expected 1`] = `
<div>
  <div
    class="outerItemListBox"
  >
    <div
      class="itemListBox"
    >
      <div
        class="innerItemListBox"
      >
        <div
          class="topicListDiv"
          id="r"
        >
          <span>
            rabbits
          </span>
        </div>
        <style>
          
            .topicListDiv {
                overflow: hidden;
                padding: var(--spacers-dp4);
                cursor: default;
            }
            .topicListDivHighlighted {
                overflow: hidden;
                padding: var(--spacers-dp4);
                background-color: var(--colors-teal700);
                cursor: default;
            }
            .topicListDivHighlighted span {
                color: var(--colors-white);
            }
            .topicListDiv:hover {
                background-color: var(--colors-grey200);
            }
        
        </style>
        <div
          class="topicListDiv"
          id="d"
        >
          <span>
            dogs
          </span>
        </div>
        <style>
          
            .topicListDiv {
                overflow: hidden;
                padding: var(--spacers-dp4);
                cursor: default;
            }
            .topicListDivHighlighted {
                overflow: hidden;
                padding: var(--spacers-dp4);
                background-color: var(--colors-teal700);
                cursor: default;
            }
            .topicListDivHighlighted span {
                color: var(--colors-white);
            }
            .topicListDiv:hover {
                background-color: var(--colors-grey200);
            }
        
        </style>
      </div>
      <div
        class="actionButtonsContainer"
      >
        <div
          class="actionButtons"
        >
          <div
            class="actionButton"
          >
            <button
              class="jsx-2371629422 small icon-only"
              data-test="dhis2-uicore-button"
              disabled=""
              name="up_button"
              type="button"
            >
              <span
                class="jsx-2371629422 button-icon"
              >
                <svg
                  fill="#404B5A"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <polygon
                    points="8 2 6.94 3.06 11.13 7.25 2 7.25 2 8.75 11.13 8.75 6.94 12.94 8 14 14 8"
                    transform="rotate(-90 8 8)"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div
            class="actionButton"
          >
            <button
              class="jsx-2371629422 small icon-only"
              data-test="dhis2-uicore-button"
              disabled=""
              name="down_button"
              type="button"
            >
              <span
                class="jsx-2371629422 button-icon"
              >
                <svg
                  fill="#404B5A"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <polygon
                    points="8 2 6.94 3.06 11.13 7.25 2 7.25 2 8.75 11.13 8.75 6.94 12.94 8 14 14 8"
                    transform="matrix(0 1 1 0 0 0)"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <style>
    
                .outerItemListBox {
                    width: 500px;
                    margin-bottom: var(--spacers-dp16);
                }
                .itemListBox {
                    margin-right: var(--spacers-dp4);
                    border: 1px solid var(--colors-grey500);
                    border-radius: 3px;
                }
                .innerItemListBox {
                    min-height: 200px;
                    max-height: 300px;
                    overflow-y: auto;
                }
                .actionButtonsContainer {
                    border-top: 1px solid var(--colors-grey500);
                    padding: var(--spacers-dp4);
                    display: flex;
                }
                .actionButtons {
                    margin-left: auto;
                    display: flex;
                }
                .actionButton {
                    margin-left: var(--spacers-dp4);
                }
            
  </style>
  <div>
    <div
      class="jsx-1141087211 "
      data-test="dhis2-uiwidgets-inputfield"
    >
      <div
        class="jsx-3375842594 "
        data-test="dhis2-uiwidgets-inputfield-content"
      >
        <div
          class="jsx-1942496442 "
          data-test="dhis2-uicore-box"
        >
          <div
            class="jsx-3353877153 jsx-1090541828 input"
            data-test="dhis2-uicore-input"
          >
            <input
              class="jsx-3353877153 jsx-1090541828 "
              placeholder="Enter a new topic name"
              type="text"
              value=""
            />
            <div
              class="jsx-3353877153 jsx-1090541828 status-icon"
            />
          </div>
        </div>
      </div>
      
    </div>
    <div
      class="addButton"
    >
      <button
        class="jsx-2371629422 primary"
        data-test="dhis2-uicore-button"
        disabled=""
        type="button"
      >
        Add Topic
      </button>
    </div>
  </div>
  <style>
    
            .addButton {
                margin-top: var(--spacers-dp4);
            }
        
  </style>
</div>
`;
